### ISSUE TITLE
fake twisted.trial._dist.test.test_workertrial.MainTestCase.fdopen returns None sometimes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6332]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#6332|
|Type|enhancement|
|Created|2013-02-25 22:28:43Z|
|Branch|https://github.com/twisted/twisted/tree/dist-fdopen-6332|

On [ticket:6314 a recent ticket related to a Python regression], this makes some of the tracebacks harder to understand, specifically, this traceback:

```
[ERROR]
Traceback (most recent call last):
  File "/var/lib/buildbot/bot-glyph-1/lucid32-py2.7maint/Twisted/twisted/plugin.py", line 181, in getCache
    dropinPath.setContent(pickle.dumps(dropinDotCache))
  File "/var/lib/buildbot/bot-glyph-1/lucid32-py2.7maint/Twisted/twisted/python/filepath.py", line 1343, in setContent
    f.close()
exceptions.AttributeError: 'NoneType' object has no attribute 'close'

twisted.trial._dist.test.test_workertrial.MainTestCase.test_empty
twisted.trial._dist.test.test_workertrial.MainTestCase.test_empty
twisted.trial._dist.test.test_workertrial.MainTestCase.test_otherReadError
twisted.trial._dist.test.test_workertrial.MainTestCase.test_otherReadError
twisted.trial._dist.test.test_workertrial.MainTestCase.test_readInterrupted
twisted.trial._dist.test.test_workertrial.MainTestCase.test_readInterrupted
```

This is a bogus error, apparently occurring because the fd value is something that we weren't expecting.  Since '3' and '4' are not standard file descriptor values (the only "special" ones are 0, 1, 2) this fixture should probably be a little more resilient.  Even if it remains as-is, it should fail earlier and in a more coherent way if a non-supported FD is passed in; it's not valid for `fdopen` to return `None`.

<details><summary>Searchable metadata</summary>

```
trac-id__6332 6332
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_dist_fdopen_6332 branches/dist-fdopen-6332
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1361831323000000 1361831323000000
changetime__1361912572000000 1361912572000000
version__None None
owner__therve therve

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [37257]) Branching to 'dist-fdopen-6332'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

Yeah that was dumb. A better version is in the branch, not faking the whole module and being a bit more clear about the intent.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_4"></a>|@tomprince set owner to @therve|
|-|-|

  1. The `fdopen` argument to `main` needs to be documented, and should perhaps be private.
  1. While reading the tests, I noticed `test_readInterrupted` still passes if the `continue` is changed to `return` in `workertrial.main`. There should be a ticket opened to deal with this.
  1. There was a [failure](https://buildbot.twistedmatrix.com/builders/lucid64-py2.6-poll/builds/2407) on the buildbot, which I can't seem to reproduce. (I guess this is unrelated)

Please commit after addressing 1+2 and running on buildbot.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set status to `closed`|
|-|-|

(In [37270]) Merge dist-fdopen-6332

Author: therve
Reviewer: tom.prince
Fixes: [#6332](https://github.com/twisted/twisted/issues/6332)

Remove patch fdopen from test_workertrial and parametrize the tested function
instead, fixing tracebacks with 2.7 maintenance branch.
